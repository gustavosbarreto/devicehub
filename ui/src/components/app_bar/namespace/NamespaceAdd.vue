<template>
  <fragment>
    <v-list-item-title>
      <v-dialog
        v-model="showAddNamespace"
        max-width="450"
        @click:outside="cancel"
      >
        <v-card>
          <v-card-title class="headline grey lighten-2 text-center">
            Enter Namespace
          </v-card-title>
          <ValidationObserver
            ref="obs"
            v-slot="{ passes }"
          >
            <v-card-text class="caption mb-0">
              <ValidationProvider
                v-slot="{ errors }"
                name="namespace"
                rules="required|rfc1123"
              >
                <v-text-field
                  v-model="namespaceName"
                  label="Namespace"
                  :error-messages="errors"
                  require
                />
              </ValidationProvider>
            </v-card-text>
            <v-card-actions>
              <v-spacer />
              <v-btn
                text
                @click="cancel"
              >
                Close
              </v-btn>
              <v-btn
                color="primary"
                text
                @click="passes(addNamespace)"
              >
                Add
              </v-btn>
            </v-card-actions>
          </ValidationObserver>
        </v-card>
      </v-dialog>
    </v-list-item-title>
  </fragment>
</template>

<script>

import {
  ValidationObserver,
  ValidationProvider,
} from 'vee-validate';

export default {
  name: 'NamespaceAdd',

  components: {
    ValidationProvider,
    ValidationObserver,
  },

  props: {
    firstNamespace: {
      type: Boolean,
      default: false,
    },

    show: {
      type: Boolean,
      required: true,
    },
  },

  data() {
    return {
      dialog: false,
      namespaceName: '',
    };
  },

  computed: {
    showAddNamespace: {
      get() {
        return this.show;
      },

      set(value) {
        this.$emit('show', value);
      },
    },
  },

  methods: {
    cancel() {
      this.dialog = false;
      this.$refs.obs.reset();
      this.namespaceName = '';
      this.$emit('update:show', false);
    },

    async switchIn(tenant) {
      try {
        await this.$store.dispatch('namespaces/switchNamespace', {
          tenant_id: tenant,
        });
        window.location.reload();
      } catch {
        this.$store.dispatch('snackbar/showSnackbarErrorLoading', this.$errors.namespaceSwitch);
      }
    },

    async addNamespace() {
      try {
        const response = await this.$store.dispatch('namespaces/post', {
          name: this.namespaceName,
        });
        if (this.$props.firstNamespace) {
          await this.switchIn(response.data.tenant_id);
        } else {
          await this.$store.dispatch('namespaces/fetch');
          this.$emit('update:show', false);
        }
        this.dialog = false;
        this.namespaceName = '';
        this.$refs.obs.reset();
        this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.namespaceCreating);
      } catch {
        this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.namespaceCreating);
      }
    },
  },
};

</script>
