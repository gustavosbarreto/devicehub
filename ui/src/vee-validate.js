import {
  required,
  integer,
  email,
  confirmed,
} from 'vee-validate/dist/rules';
import { extend } from 'vee-validate';
import isValidHostname from 'is-valid-hostname';

import { parsePrivateKey, parseKey } from '@/sshpk';

extend('required', {
  ...required,
  message: 'This field is required',
});

extend('integer', {
  ...integer,
  message: 'This value must be a integer number',
});

extend('email', {
  ...email,
  message: 'This field must be a valid email',
});

extend('noDot', {
  validate: (value) => !/\./.test(value),
  message: 'The name must not contain dots',
});

extend('rfc1123', {
  validate: (value) => isValidHostname(value),
  message: 'You entered an invalid RFC1123 name',
});

extend('routeIdentifier', {
  validate: (value) => !/\/|@|&|:/.test(value),
  message: 'The name must not contain /, @, &, and :.',
});

extend('password', (value) => {
  if (value.length < 5 || value.length > 30) {
    return 'Your password should be 5-30 characters long';
  }
  return true;
});

extend('namespace', (value) => {
  if (value.length < 3 || value.length > 30) {
    return 'Your namespace should be 3-30 characters long';
  }
  return true;
});

extend('device', (value) => {
  if (value.length < 3 || value.length > 30) {
    return 'Your hostname should be 3-30 characters long';
  }
  return true;
});

extend('tag', (value) => {
  if (value.length < 3 || value.length > 255) {
    return 'Your tag should be 3-255 characters long';
  }
  return true;
});

extend('tagRequired', {
  ...required,
  message: 'You must choose at least one tag',
});

extend('tagsLength', {
  validate(value) {
    if (value.length > 3) {
      return 'The maximum capacity has reached';
    }
    return true;
  },
});

extend('comparePasswords', {
  validate(value, { currentPassword }) {
    if (value === currentPassword) {
      return false;
    }
    return true;
  },
  params: ['currentPassword'],
  message: 'The passwords are the same',
});

extend('confirmed', {
  ...confirmed,
  message: 'The passwords do not match',
});

extend('parseKey', {
  validate(value, args) {
    try {
      if (args.typeKey === 'private') {
        parsePrivateKey(value);
      } else {
        parseKey(value);
      }

      return true;
    } catch (err) {
      return false;
    }
  },
  params: ['typeKey'],
  message: 'Not valid key',
});
