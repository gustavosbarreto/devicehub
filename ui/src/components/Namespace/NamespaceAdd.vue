<template>
  <div class="ml-2 mr-2">
    <v-btn
      v-if="!firstNamespace"
      block
      :size="isSmall ? 'small' : 'default'"
      color="primary"
      @click="showDialog = true"
    >
      Add Namespace
    </v-btn>

    <v-list-item-title>
      <v-dialog
        v-model="showDialog"
        @click:outside="update"
        :max-width="!openVersion ? '400' : '650'"
        v-bind="$attrs"
      >
        <v-card data-test="namespaceAdd-card" class="bg-v-theme-surface">
          <div v-if="!openVersion">
            <v-card-title class="text-headline bg-primary">
              Enter Namespace
            </v-card-title>

            <v-card-text>
              <v-text-field
                v-model="namespaceName"
                label="Username"
                :error-messages="namespaceNameError"
                required
                variant="underlined"
                data-test="username-text"
              />
            </v-card-text>

            <v-card-actions>
              <v-spacer />
              <v-btn data-test="close-btn" @click="update"> Close </v-btn>

              <v-btn color="primary" data-test="add-btn" @click="addNamespace">
                Add
              </v-btn>
            </v-card-actions>
          </div>
          <div v-else>
            <v-card-title class="bg-primary">
              Add a namespace using the CLI
            </v-card-title>

            <v-card-text class="mt-4 mb-0 pb-1 mb-4">
              <p class="text-body-2">
                In the Community Edition of ShellHub, namespaces must be added using the administration CLI.
                For detailed instructions on how to add namespaces, please refer to the documentation at the ShellHub Administration Guide.
              </p>
              <div id="cli-instructions" class="mt-3 text-body-2">
                <p class="text-caption mb-0 mt-3" data-test="openContentSecond-text">
                  Check the
                  <a
                    :href="'https://docs.shellhub.io/self-hosted/administration'"
                    target="_blank"
                    rel="noopener noreferrer"
                  >ShellHub Administration Guide</a
                  >
                  for more information.
                </p>
              </div>
            </v-card-text>
          </div>
        </v-card>
      </v-dialog>
    </v-list-item-title>
  </div>
</template>

<script setup lang="ts">
import { computed, ref } from "vue";
import * as yup from "yup";
import { useField } from "vee-validate";
import axios, { AxiosError } from "axios";
import {
  INotificationsError,
  INotificationsSuccess,
} from "../../interfaces/INotifications";
import { useStore } from "../../store";
import handleError from "../../utils/handleError";
import { envVariables } from "@/envVariables";

const props = defineProps({
  firstNamespace: {
    type: Boolean,
    default: false,
  },
  show: {
    type: Boolean,
    default: false,
  },
  isSmall: {
    type: Boolean,
    default: false,
  },
  enableSwitchIn: {
    type: Boolean,
    default: false,
  },
});
const emit = defineEmits(["update"]);
const store = useStore();

const showDialog = ref(false);

const {
  value: namespaceName,
  errorMessage: namespaceNameError,
  setErrors: setNamespaceNameError,
  resetField: resetNamespaceName,
} = useField<string>(
  "namespaceName",
  yup
    .string()
    .required()
    .min(3)
    .max(30)
    .matches(/^[^.]*$/, "The name must not contain dots"),
  {
    initialValue: "",
  },
);

const switchIn = async (tenant: string) => {
  try {
    await store.dispatch("namespaces/switchNamespace", {
      tenant_id: tenant,
    });
    window.location.reload();
  } catch (error: unknown) {
    store.dispatch(
      "snackbar/showSnackbarErrorLoading",
      INotificationsError.namespaceSwitch,
    );
    handleError(error);
  }
};

const close = () => {
  showDialog.value = false;
  resetNamespaceName();
};

const update = () => {
  emit("update");
  close();
};

const addNamespace = async () => {
  if (!namespaceNameError.value) {
    try {
      // const tenant = localStorage.getItem("tenant");

      const response = await store.dispatch("namespaces/post", namespaceName.value);

      if (props.firstNamespace || props.enableSwitchIn) {
        await switchIn(response.data.tenant_id);
        close();
      } else {
        await store.dispatch("namespaces/fetch", {
          page: 1,
          perPage: 30,
        });
        update();
      }

      store.dispatch(
        "snackbar/showSnackbarSuccessAction",
        INotificationsSuccess.namespaceCreating,
      );
    } catch (error: unknown) {
      if (axios.isAxiosError(error)) {
        const axiosError = error as AxiosError;
        if (axiosError.response?.status === 400) {
          setNamespaceNameError(
            "Your namespace should be 3-30 characters long",
          );
        } else if (axiosError.response?.status === 403) {
          setNamespaceNameError("Update your plan to create more namespaces");
        } else if (axiosError.response?.status === 409) {
          setNamespaceNameError("namespace already exists");
        }
      } else {
        store.dispatch(
          "snackbar/showSnackbarErrorAction",
          INotificationsError.namespaceCreating,
        );
        handleError(error);
      }
    }
  }
};

const openVersion = computed(() => !envVariables.isCloud || !envVariables.isEnterprise);
</script>
