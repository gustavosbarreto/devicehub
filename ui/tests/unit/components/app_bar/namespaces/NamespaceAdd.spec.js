import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import { ValidationProvider, ValidationObserver } from 'vee-validate';
import flushPromises from 'flush-promises';
import Vuetify from 'vuetify';
import NamespaceAdd from '@/components/app_bar/namespace/NamespaceAdd';
import '@/vee-validate';

describe('NamespaceAdd', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();
  localVue.use(Vuex);
  localVue.component('ValidationProvider', ValidationProvider);
  localVue.component('ValidationObserver', ValidationObserver);

  document.body.setAttribute('data-app', true);

  let wrapper;

  const show = true;
  const firstNamespace = true;

  const invalidNamespaces = [
    '\'', '"', '!', '@', '#', '$', '%', '¨', '&', '*', '(', ')', '-', '_', '=', '+', '´', '`', '[',
    '{', '~', '^', ']', ',', '<', '..', '>', ';', ':', '/', '?',
  ];

  const invalidMinAndMaxCharacters = [
    's', 'sh', 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx', 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',
  ];

  const store = new Vuex.Store({
    namespaced: true,
    state: {
    },
    getters: {
    },
    actions: {
      'namespaces/switchNamespace': () => {},
      'namespaces/post': () => {},
      'namespaces/fetch': () => {},
      'snackbar/showSnackbarSuccessAction': () => {},
      'snackbar/showSnackbarErrorAction': () => {},
      'snackbar/showSnackbarErrorLoading': () => {},
    },
  });

  ///////
  // In this case, the rendering of the dialog is checked. In which
  // case with the input data it cannot take place.
  ///////

  describe('Doesn\'t render the dialog', () => {
    beforeEach(() => {
      wrapper = mount(NamespaceAdd, {
        store,
        localVue,
        stubs: ['fragment'],
        propsData: { show: !show, firstNamespace },
        vuetify,
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Receive data in props', () => {
      expect(wrapper.vm.show).toEqual(!show);
      expect(wrapper.vm.firstNamespace).toEqual(firstNamespace);
    });
    it('Compare data with default value', () => {
      expect(wrapper.vm.dialog).toEqual(false);
      expect(wrapper.vm.namespaceName).toEqual('');
    });
    it('Process data in the computed', () => {
      expect(wrapper.vm.showAddNamespace).toEqual(false);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', () => {
      expect(wrapper.find('[data-test="namespaceAdd-card"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="namespace-text"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="close-btn"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="add-btn"]').exists()).toBe(false);
    });
  });

  ///////
  // In this case, the rendering of the dialog is checked. In which
  // case with the input data it cannot take place.
  ///////

  describe('Dialog', () => {
    beforeEach(() => {
      wrapper = mount(NamespaceAdd, {
        store,
        localVue,
        stubs: ['fragment'],
        propsData: { show, firstNamespace },
        vuetify,
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Receive data in props', () => {
      expect(wrapper.vm.show).toEqual(show);
      expect(wrapper.vm.firstNamespace).toEqual(firstNamespace);
    });
    it('Compare data with default value', () => {
      expect(wrapper.vm.dialog).toEqual(false);
      expect(wrapper.vm.namespaceName).toEqual('');
    });
    it('Process data in the computed', () => {
      expect(wrapper.vm.showAddNamespace).toEqual(true);
    });

    //////
    // HTML validation
    //////

    //////
    // In this case, the empty fields are validated.
    //////

    it('Show validation messages', async () => {
      wrapper.setData({ namespaceName: '' });
      await flushPromises();

      const validator = wrapper.vm.$refs.providerNamespace;

      await validator.validate();
      expect(validator.errors[0]).toBe('This field is required');
    });

    //////
    // In this case, the error for dot are validated.
    //////

    it('Show validation messages', async () => {
      wrapper.setData({ namespaceName: 'ShelHub.' });
      await flushPromises();

      const validator = wrapper.vm.$refs.providerNamespace;

      await validator.validate();
      expect(validator.errors[0]).toBe('The name must not contain dots');
    });

    //////
    // In this case, the RFC1223 are validated.
    //////

    it('Show validation messages', async (done) => {
      invalidNamespaces.forEach(async (inamespace) => {
        wrapper.setData({ namespaceName: inamespace });
        await flushPromises();

        const validator = wrapper.vm.$refs.providerNamespace;

        await validator.validate();
        expect(validator.errors[0]).toBe('You entered an invalid RFC1123 name');

        await flushPromises();
        done();
      });
    });

    //////
    // In this case, the min and max characters are validated.
    //////

    it('Show validation messages', async (done) => {
      invalidMinAndMaxCharacters.forEach(async (character) => {
        wrapper.setData({ namespaceName: character });
        await flushPromises();

        const validator = wrapper.vm.$refs.providerNamespace;

        await validator.validate();
        expect(validator.errors[0]).toBe('Your namespace should be 3-30 characters long');

        await flushPromises();
        done();
      });
    });
    it('Renders the template with data', () => {
      expect(wrapper.find('[data-test="namespaceAdd-card"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="namespace-text"]').element.value).toEqual('');
      expect(wrapper.find('[data-test="close-btn"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="add-btn"]').exists()).toBe(true);
    });
  });
});
