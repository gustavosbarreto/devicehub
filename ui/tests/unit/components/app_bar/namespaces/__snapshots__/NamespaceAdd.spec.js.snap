// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NamespaceAdd Renders the component 1`] = `
<fragment-stub>
  <div class="v-list-item__title">
    <div role="dialog" class="v-dialog__container"></div>
  </div>
</fragment-stub>
`;
